### ISSUE TITLE
Multithreading

### ISSUE THREAD

**Issue by mdestefano** [ishepard/pydriller#158]
I used to exploit [Repodriller](https://github.com/mauricioaniche/repodriller), but since it is no longer maintained and pydriller presents a lot of useful functionalities for researchers (eg. list of modified methods), I switched to it.
However, currently, repository mining can happen only in sequential, and running external tools on a checkout takes an impossible time. It would be nice to have a concurrent analysis like the one that was implemented in Repodriller.
Thanks a lot, and thanks for the great job done so far!

**Comment by JulianVolodia** [comment]
I think we should co-work on this and https://github.com/ishepard/pydriller/issues/102 because ofc features are important but this module seamlessly use memory harsher than Java does ;P and even not limit itself if that could cause problems (just fall from bike and is happy C: )

**Comment by ishepard** [comment]
Hi! Sorry for the late reply, I've been busy at work.
Multithreading is something that we had in mind for a while, but implementing right is quite challenging, for multiple reasons:
- python is not a language that likes it very much
- git doesn't allow it
- bad things can happen when you multithreading

For example, how can we ensure that users don't use `git checkout` when having multiple threads? 
I'm not saying that I am against multithreading, but it's just a lot of work for (what I think will happen) a very small gain. After all, even git doesn't perform many operations on multiple processes/threads, why should Pydriller? 😄 
We can probably try to read from a repo using different threads, but we need to make sure that the behaviour remains correct.

What I think could be more valuable is to run separate processes for each repository: so if you pass more than 1 repo, we will run each study in a different thread.

Again, git doesn't like to be threaded..so we need to check things are normal.

**Comment by marco-c** [comment]
I've been using pydriller in multiple threads for a while, and it's worked well enough for my purposes (I can't be sure it will work for all use cases).
There is some discussion in #110, in particular see the example in https://github.com/ishepard/pydriller/issues/110#issuecomment-605092557.

@ishepard it *shouldn't* be a problem for git because pydriller is not using libgit directly but running the `git` executable.

**Comment by ishepard** [comment]
@marco-c how did you multithreaded pydriller? In terms of commits (like if we have 100 commits to analyze, we split them in 5 threads), or in terms of repos (we have to analize 2 repos, so we use 2 threads)?

**Comment by marco-c** [comment]
> @marco-c how did you multithreaded pydriller? In terms of commits (we have 100 commits to analyze, we split them in 5 threads, 20 commits each), or in terms of repos (we have to analize 2 repos, so we use 2 threads)?

In terms of commits, similarly to https://github.com/ishepard/pydriller/issues/110#issuecomment-605092557

**Comment by ishepard** [comment]
I see. 
So from what I see, when we use threading in RepositoryMining we need to lock the creation of GitRepository, while everything else should pretty much work as is. 

I can try to have a look at it this week if I have time. 

I was also considering to write the entire Pydriller using async/await, maybe it will improve something.

**Comment by JulianVolodia** [comment]
maybe refactor would be good option, but maybe we should try mitigate problems first?

May I help You in any way? Maybe we could do some option to use queue(s) (the best 2, for heavy and light tasks) ?

I use pydriller right away from django and probably would introduce celery queues to my project anyway. Chopping the responsibilities and trying to use lighter thing (should libgit be lighter? I havent use libgit @marco-c ...

Async would be nice but still there is other problem imho. @marco-c You said that pydriller use git executable, but is git *awaiting*? Afaik no. So even @ishepard rewrite the Pydriller to async there would be i/o waiting? If we await for heavy process like git this would be (maybe I am narrow minded and dont get idea - sorry) catastrophe, so we will run like 1M green threads because all awaited on git return value, in my opinion it would look like fork bomb...

Again, sorry for interrupting, if I could do sth, test or write - please tell me. Appreciate the possibility to contribute!

Kind regards, stay safe 👋

**Comment by marco-c** [comment]
> I was also considering to write the entire Pydriller using async/await, maybe it will improve something.

I think using async/await would have basically the same effects of multithreading in this case, given the way multithreading works in Python (the actual parallelism only occurs when a thread is not executing Python code but waiting on I/O).
With async, whenever you run `git`, you will await and let another async task run.
With threads, whenever you run `git`, you will "pause" the thread and let another thread run.

The only drawback with async is that you need to be careful and avoid using functions that perform "normal" (i.e. non-async) I/O. E.g. if you run "os.mkdir" in an async task, it will block *all* tasks. If you run "os.mkdir" in a thread, it will only block the thread that called it and let other threads run.
This means converting to a async-based implementation can be more work than converting to a multithreading-based implementation.

**Comment by mdestefano** [comment]
Hi all! Thanks for your replies.
I looked into Repodriller and I found out that when using multithreading and visitors do change the status of the repository, it creates multiple copies of the repositories (don't know if they're multiple git clone operations or just copies), and it gives a copy for each thread. In this case, there are no race conditions. Would it be possible to implement a similar solution to enable multithreading in pydriller?

**Comment by ishepard** [comment]
@mdestefano it might be useful in case we would like to allow users to perform git checkout when multithreading. Otherwise I'd say it's irrelevant, git operations will be performed in isolation. Let's keep it in mind though!

**Comment by ishepard** [comment]
First analysis with multithreading:
From 11.04 minutes to 1.46 😨

**Comment by mdestefano** [comment]
Wow! Such an improvement!  👏🏼 👏🏼 👏🏼 
So, are you going to implement it in production?

**Comment by ishepard** [comment]
I'll do it soon. This is gonna be part of a bigger update for Pydriller 2.0. 
I'm currently waiting for inspirations/ideas on what could be included. See #159.
If I can't think of any other refactoring, I will publish it soon.

**Comment by ishepard** [comment]
Closing this issue since I solved it in branch pydriller2.0! Thanks everyone!
